### ISSUE TITLE
Aggregate queries with case/when statement fails for SQL Server / sp_prepexec with differing parameters in select vs group by portion of query

### ISSUE THREAD

**Issue by dinabilenkis** [mkleehammer/pyodbc#479]
#. issue 
Using: "get_queryset" in django for an aggregate query that includes case/when and sql_server_pyodbc, the query will fail.

This error happens because of sp_prepexec contains a different set of parameters in the select and the group by sections of query. (even though the values are the same)

an example query sent to sql server:
`select case when col1=@p1 then @p2 end as new_col1, sum(col2) from table group by case when col1=@p3 then @p4 end`

This will fail because the case/when in the select must be the same as in the group by section. 
The query generated should be like this (no longer using @p3 and @p4 parameters.

`select case when col1=@p1 then @p2 end as new_col1, sum(col2) from table group by case when col1=@p1 then @p2 end`

#. environment 
- pyodbc: 4.0.24
- OS: Ubuntu 16.04
- DB: SQL Server 2016/2017
- driver: SQL Server ODBC for Linux 17

**Comment by v-chojas** [comment]
Could you post an [ODBC trace](https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/data-access-tracing-with-the-odbc-driver-on-linux) ? pyODBC does not do any query processing itself, and only passes queries from the application through to the driver.

**Comment by gordthompson** [comment]
I don't believe that this is a problem with pyodbc *per se*. It is an issue with the parameterized SELECT query that is being provided to pyodbc, either as written directly by you or constructed by the Django database backend you are using.

This doesn't work using vanilla Python_3/pyodbc

```python
sql = """\
select case when col1=? then ? end as new_col1, sum(col2) 
from table1 group by case when col1=? then ? end
"""
params = ('Marco', 'Polo', 'Marco', 'Polo')
print(crsr.execute(sql, params).fetchall())
```

failing with

> [42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Column 'table1.col1' is invalid in the select list because it is not contained in either an aggregate function or the GROUP BY clause. (8120) (SQLExecDirectW)"

However, this works fine

```python
sql = """\
SELECT new_col1, SUM(col2)
FROM 
(
    SELECT CASE WHEN col1=? THEN ? END AS new_col1, col2
    from table1
) AS subq
GROUP BY new_col1;
"""
params = ('Marco', 'Polo')
```

**Comment by guruofgentoo** [comment]
I would propose that pyodbc could provide the option to substitute parameter values into the query, in order to be running a fully-rendered query against SQL Server through the ODBC layer.
- With the option turned on, this would eliminate the need to prepare the statement and bind parameters, and would bridge the gap between the two existing execution methods (prepare/bind and raw pass-through execution)
- IMO the first query above from @gordthompson is valid SQL, and so there should be an option to execute it successfully
- Compatibility with SQLAlchemy would be increased
  - Other DBAPI solutions insert properly-quoted parameters as a matter of course
    - see pymssql, https://github.com/pymssql/pymssql/blob/master/src/_mssql.pyx#L1867
  - Since we can't name parameters, fixing this at the SQLAlchemy layer would require making that library the final renderer of the query, which would eliminate any potential downstream optimization for parameterized queries
    - see the discussion with Mike Bayer at https://github.com/sqlalchemy/sqlalchemy/issues/4540
- Making this an option to turn on in the connection would allow developers using pyodbc to make an informed decision as to whether to have pyodbc:
  - a) prepare statements, accepting the limitations such as the one described in this issue
  - b) insert properly-quoted parameters into the query, and execute the raw SQL

@mkleehammer would you be open to either adding this sort of option in or accepting a PR that does? My team is evaluating a transition to pyodbc of a large SQL Server application due to this library being the Microsoft recommended solution, and this is a blocking issue for us.

I appreciate your thoughts and those of the other contributors. Thanks!

**Comment by mkleehammer** [comment]
I really don't want to start modifying SQL - that's an entirely different game.  The example SQL would work if it the parameters were substituted, but I'm not sure that makes it "correct" when using parameters.

This seems like something that could trivially be added on top by wrapping an object around Cursors that preprocesses execute (and executemany?) calls.  (Note that the trivial part is intercepting.  Getting the escaping right might not be and may depend on the DB may be more difficult and risky.)

**Comment by v-chojas** [comment]
I agree with @mkleehammer --- pyODBC shouldn't try to parse SQL, given that it's intended as a generic Python-ODBC layer. In order to correctly perform parameter substitution, you _will_ need to parse SQL, and although there is a standard, there are differences between different DBs.

**Comment by hongshi-hunterfox** [comment]
```
sql = """
select case when col1=? then ? end as new_col1, sum(col2) 
from table1
group by case when col1=@p1 then @p2 end
"""
params = ('Marco', 'Polo', 'Marco', 'Polo')
```
